### PR TITLE
Respect JVM http proxy settings when using Feign.

### LINKED ISSUE #226: Respect JVM-wide proxy settings in the launcher
Otherwise okhttp requests don't get proxied, so the ping loop and subsequent post of config/files/jars from submitter to driver pod's rest service don't make it out the proxy.  Apparently okhttp doesn't respect these by default.

See https://github.com/palantir/spark/compare/master...respect-http-proxy-palantir12 which we had to do to get an internal application working.

cc @mccheah

### PR DESCRIPTION
**Pull request by mccheah**
Closes #226

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 1]
```diff
@@ -42,6 +48,49 @@ private[spark] object HttpClientUtil {
...
+                matchingUriExists(uri, httpUris)
+              }
+            case "https" =>
+              logInfo(s"Looking up https proxies to route $uri")
+              maybeHttpsProxy.filter { _ =>
+                matchingUriExists(uri, httpsUris)
+              }
+            case _ => None
+          }).getOrElse(directProxy)
+          logInfo(s"Routing $uri through ${resolvedProxy.address()} with proxy" +
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala]: all these `logInfo` should probably be at `logDebug` instead

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 2]
```diff
@@ -66,4 +115,17 @@ private[spark] object HttpClientUtil {
       .retryer(target)
       .target(target)
   }
+
+  private def matchingUriExists(uri: URI, httpUris: Set[URI]) = {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala]: put return type in signature here

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 3]
```diff
@@ -42,6 +48,49 @@ private[spark] object HttpClientUtil {
...
+            case "https" =>
+              logInfo(s"Looking up https proxies to route $uri")
+              maybeHttpsProxy.filter { _ =>
+                matchingUriExists(uri, httpsUris)
+              }
+            case _ => None
+          }).getOrElse(directProxy)
+          logInfo(s"Routing $uri through ${resolvedProxy.address()} with proxy" +
+                  s" type ${resolvedProxy.`type`()}")
+          Collections.singletonList(resolvedProxy)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala]: is the indentation here a little off?  83-85 should be 2 spaces leftward?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 4]
```diff
@@ -42,6 +48,49 @@ private[spark] object HttpClientUtil {
...
+    val maybeHttpProxy = Option.apply(System.getProperty(Config.KUBERNETES_HTTP_PROXY))
+      .orElse(maybeAllProxy)
+      .map(uriStringToProxy)
+    val maybeHttpsProxy = Option.apply(System.getProperty(Config.KUBERNETES_HTTPS_PROXY))
+      .orElse(maybeAllProxy)
+      .map(uriStringToProxy)
+    val maybeNoProxy = Option.apply(System.getProperty(Config.KUBERNETES_NO_PROXY))
+        .map(_.split(","))
+        .toSeq
+        .flatten
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala]: check indentation -- 2 spaces

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala [hunk 5]
```diff
@@ -42,6 +48,49 @@ private[spark] object HttpClientUtil {
     Option.apply(trustContext).foreach(context => {
       httpClientBuilder = httpClientBuilder.sslSocketFactory(sslSocketFactory, context)
     })
+    val uriObjects = uris.map(URI.create)
+    val httpUris = uriObjects.filter(uri => uri.getScheme == "http")
+    val httpsUris = uriObjects.filter(uri => uri.getScheme == "https")
+    val maybeAllProxy = Option.apply(System.getProperty(Config.KUBERNETES_ALL_PROXY))
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala:54]: can these system property manipulations be moved out of this method?  We probably don't need to run those in every client setup (unless we want to late bind to the values of the properties at that time)

**Review by ash211 — COMMENTED** [review]
Only minor nits

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala (hunk 5 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala:54]: We should keep it at this level to ensure that we're always picking up the properties whenever we set up HTTP clients. In other words, this class is opinionated towards using the system properties to set up the proxy. If we wanted to allow for multiple opinions, for example also allowing the proxy configuration to be set via `SparkConf`, then we should inject these values from the caller.

**Comment by mccheah** [comment]
rerun unit tests please

**Comment by ash211** [comment]
rerun unit tests please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala (hunk 5 above)
Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala:54]: Ok that makes sense -- want to grab the system properties at the time the client is created, not the first time the class is accessed (which is less predictable)

**Review by ash211 — APPROVED** [review]
LGTM and tested in an internal app

@foxish a final look please before merging?

**Comment by foxish** [comment]
LGTM, this is good. I think @iyanuobidele had a use case similar to this.

**Pull request merged**